### PR TITLE
Added autoDelegations management

### PR DESCRIPTION
**Pull request by LotuxPunk**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,7 +17,8 @@
     "prepare": "rimraf dist && npm run build && jq '{name:.name, version:.version, description:.description, main:\"index.js\", types:\"index.d.ts\", dependencies:.dependencies, author:.author, license:.license, bugs:.bugs, homepage:.homepage}' < package.json > dist/package.json",
     "publish": "yarn version && yarn run prepare && cd dist && yarn publish --no-git-tag-version --new-version `git describe --abbrev=0 --tags` && cd ..",
     "precommit": "pretty-quick --staged",
-    "test": "nyc mocha test/**/*.ts --require ts-node/register --require source-map-support/register --timeout=60000 --recursive"
+    "test": "nyc mocha test/**/*.ts --require ts-node/register --require source-map-support/register --timeout=60000 --recursive",
+    "test-user": "nyc mocha test/apis/user-api.ts --require ts-node/register --require source-map-support/register --timeout=60000 --recursive"
```
Comment (LotuxPunk) [inline package.json:21]:
```suggestion

```

File: src/apis/UserApi.ts [hunk 2]
```diff
@@ -94,4 +94,22 @@ export interface UserApi {
...
   ): Promise<Connection>
+
+  /**
+   * Add autoDelegations values to the user.
+   * @param type Type of AutoDelegation to add
+   * @param user User to update
+   * @param to Array of DataOwnerId to add
+   * @return Updated user
+   */
+  addAutoDelegationsTo(type: DelegationTag, user: User, to: string[]): Promise<User>
```
Comment (lisatr2812) [inline src/apis/UserApi.ts:105]: Rename it to `shareAllFutureDataWith(sharedDataType: SharedDataType, dataOwnerIds: string[])`

Comment (lisatr2812) [inline src/apis/UserApi.ts:105]: `user` does not need to be used in this method : The service can be called only for the current user. 
Meaning user may only be the usserApi.getCurrentUser(), which you can call directly in the service

File: src/apis/UserApi.ts [hunk 3]
```diff
@@ -94,4 +94,22 @@ export interface UserApi {
...
+
+  /**
+   * Removes autoDelegations values to the user.
+   * @param type Type of AutoDelegation to removes
+   * @param user User to update
+   * @param to Array of DataOwnerId to add
+   * @return Updated user
+   */
+  removeAutoDelegationsTo(type: DelegationTag, user: User, to: string[]): Promise<User>
```
Comment (lisatr2812) [inline src/apis/UserApi.ts:114]: `user` does not need to be used in this method : The service can be called only for the current user. 
Meaning user may only be the usserApi.getCurrentUser(), which you can call directly in the service

Comment (lisatr2812) [inline src/apis/UserApi.ts:114]: Rename it to `stopSharingDataWith(sharedDataType: SharedDataType, dataOwnerIds: string[])`

File: src/apis/impl/UserApiImpl.ts [hunk 4]
```diff
@@ -160,10 +160,59 @@ export class UserApiImpl implements UserApi {
...
-    options: {keepAlive?: number, lifetime?: number, connectionMaxRetry?: number, connectionRetryIntervalMs?: number } = {}
+    options: { keepAlive?: number, lifetime?: number, connectionMaxRetry?: number, connectionRetryIntervalMs?: number } = {}
   ): Promise<Connection> {
     return subscribeToEntityEvents(this.basePath, this.username!, this.password!, "User", eventTypes, filter, eventFired, options)
       .then((rs) => new ConnectionImpl(rs))
   }
 
+  async addAutoDelegationsTo(type: DelegationTag, user: User, to: string[]): Promise<User> {
+    let newAutoDelegations
+    if (user.autoDelegations[type]) {
```
Comment (lisatr2812) [inline src/apis/impl/UserApiImpl.ts:171]: `const user = this.userApi.getCurrentUser()`

File: src/apis/impl/UserApiImpl.ts [hunk 5]
```diff
@@ -160,10 +160,59 @@ export class UserApiImpl implements UserApi {
...
+      newAutoDelegations = Object.entries(user.autoDelegations).reduce((accumulator, [key, values]) => {
```
Comment (lisatr2812) [inline src/apis/impl/UserApiImpl.ts:172]: Rename `newAutoDelegations` to `newDataSharing`

File: src/apis/impl/UserApiImpl.ts [hunk 6]
```diff
@@ -160,10 +160,59 @@ export class UserApiImpl implements UserApi {
...
+          return {...accumulator, [key]: [...values]}
+        }, {}),
+        [type]: to
+      }
+    }
+
+    const updatedUserDto = await this.userApi.modifyUser(
+      UserMapper.toUserDto({
+        ...user,
+        autoDelegations: newAutoDelegations as { [key in DelegationTag]: Set<string> }
```
Comment (lisatr2812) [inline src/apis/impl/UserApiImpl.ts:187]: Rename user property `autoDelegations` to `sharingDataWith`

File: src/apis/impl/UserApiImpl.ts [hunk 7]
```diff
@@ -160,10 +160,59 @@ export class UserApiImpl implements UserApi {
...
+
+    if (updatedUserDto != undefined) {
+      return UserMapper.toUser(updatedUserDto)!
+    }
+
+    throw new Error("Couldn't add auto-delegations to user")
+  }
+
+  async removeAutoDelegationsTo(type: DelegationTag, user: User, to: string[]): Promise<User> {
+    const newAutoDelegations = Object.entries(user.autoDelegations).reduce((accumulator, [key, values]) => {
```
Comment (lisatr2812) [inline src/apis/impl/UserApiImpl.ts:199]: Rename `newAutoDelegations` to `newDataSharing`

File: src/mappers/user.ts [hunk 8]
```diff
@@ -15,7 +15,7 @@ export namespace UserMapper {
         id: obj.id,
         properties: new Set(map(obj.properties, toProperty)),
         roles: new Set(obj.roles),
-        autoDelegations: toMapSet(obj.autoDelegations),
+        autoDelegations: toMapSet(obj.autoDelegations) as { [key in DelegationTag]: Set<string>; } | undefined,
```
Comment (lisatr2812) [inline src/mappers/user.ts:18]: Rename user property (only in MedTech !) `autoDelegations` to `sharingDataWith`

File: src/models/User.ts [hunk 9]
```diff
@@ -75,28 +75,50 @@ constructor(json: IUser) {
...
     /**
-    * Id of the patient if the user is a patient
-    */
+     * Id of the patient if the user is a patient
+     */
     'deviceId'?: string;
+  /**
+   * Delegations that are automatically generated client side when a new database object is created by this user
+   */
+  'autoDelegations': { [key in DelegationTag]: Set<string>; };
```
Comment (lisatr2812) [inline src/models/User.ts:84]: Rename user property (only in MedTech !) `autoDelegations` to `sharingDataWith`

File: src/models/User.ts [hunk 10]
```diff
@@ -75,28 +75,50 @@ constructor(json: IUser) {
     */
     'patientId'?: string;
     /**
-    * Id of the patient if the user is a patient
-    */
+     * Id of the patient if the user is a patient
+     */
     'deviceId'?: string;
+  /**
+   * Delegations that are automatically generated client side when a new database object is created by this user
```
Comment (lisatr2812) [inline src/models/User.ts:82]: `Ids of the dataOwners with who the user is sharing all new data he is creating in iCure : All Data Types that may be shared can be found in SharedDataType enum`

File: src/models/User.ts (hunk 9 above)
Comment (lisatr2812) [inline src/models/User.ts:84]: Rename `DelegationTag` to `SharedDataType`

File: src/models/User.ts [hunk 11]
```diff
@@ -75,28 +75,50 @@ constructor(json: IUser) {
...
-    */
-    'authenticationTokens': { [key: string]: AuthenticationToken; };
+  /**
+   * Encrypted and time-limited Authentication tokens used for inter-applications authentication
+   */
+  'authenticationTokens': { [key: string]: AuthenticationToken; };
 
 }
 
+export type DelegationTag =
```
Comment (lisatr2812) [inline src/models/User.ts:100]: Rename `DelegationTag` to `SharedDataType`

Comment (lisatr2812) [inline src/models/User.ts:100]: In this enum, let's only use the following ones : 
```
all,
administrativeData,
generalInformation,
financialInformation,
medicalInformation,
sensitiveInformation,
confidentialInformation
```
The other values are very specific to Belgium codes, let's try to not expose it first. We can expose it later if needed

File: test/apis/hcp-user.ts [hunk 12]
```diff
@@ -7,6 +7,14 @@ import {webcrypto} from "crypto";
 import {User} from "../../src/models/User";
 import {HealthcareProfessional} from "../../src/models/HealthcareProfessional";
 import {SystemMetaDataOwner} from "../../src/models/SystemMetaDataOwner";
+import {TestUtils} from "../test-utils";
+import * as os from "os";
+import {LocalStorage} from "node-localstorage";
+
+const tmp = os.tmpdir();
```
Comment (lisatr2812) [inline test/apis/hcp-user.ts:14]: Use the `setLocalStorage(fetch)` method available in test-utils

File: test/apis/user-api.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
+import {getEnvVariables, TestUtils} from "../test-utils";
+import {v4 as uuid} from 'uuid';
+import {assert} from "chai";
+import 'isomorphic-fetch'
+import {tmpdir} from "os";
+import {TextDecoder, TextEncoder} from "util";
+
+(global as any).localStorage = new (require('node-localstorage').LocalStorage)(tmpdir(), 5 * 1024 * 1024 * 1024)
```
Comment (lisatr2812) [inline test/apis/user-api.ts:8]: Use the setLocalStorage(fetch) method available in test-utils

File: test/apis/user-api.ts [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+;(global as any).fetch = fetch
+;(global as any).Storage = ''
+;(global as any).TextDecoder = TextDecoder
+;(global as any).TextEncoder = TextEncoder
+
+const {iCureUrl: iCureUrl, msgGtwUrl: msgGtwUrl, authProcessHcpId: authProcessHcpId, specId: specId} = getEnvVariables()
+
+describe('User API', () => {
+  it('Management of auto-delegations', async () => {
```
Comment (lisatr2812) [inline test/apis/user-api.ts:17]: Rename your test to `a user is able to share data with another dataOwner, and stop sharing data with him later`

Add the following test as well : 
- `When a user already shares data with the provided dataOwner, the user is returned successfully, without any additional request to iCure` and check that the user is exactly the same than the provided one

**Review by lisatr2812 — CHANGES_REQUESTED** [review]

**Pull request merged**